### PR TITLE
CWL Tool Descriptions (& other work)

### PR DESCRIPTION
**Pull request by douglowe**
https://douglowe.github.io/cwl-novice-tutorial/

This PR includes a few changes:

- removal of some episodes which have moved to the intermediate lesson
- added some setup text for installing cwl
- started the introductionary material
- code blocks and Benten diagrams added for teaching about Tool Descriptions and a single step Workflow

The Tool Description material is currently written as a simple 'hello world' example - we should replace this with a similar level complexity domain example, but the general layout of the lesson itself I think can remain the same.

### CODE REVIEW

**Comment by mr-c** [comment]
@douglowe Shall I review your PR during this afternoon's monthly curriculum meetup?

**Comment by douglowe** [comment]
> @douglowe Shall I review your PR during this afternoon's monthly curriculum meetup?

that would be good - I'll pop online now.

File: _episodes/01-introduction.md [hunk 1]
```diff
@@ -12,7 +12,18 @@ objectives:
 keypoints:
 - "First key point. Brief Answer to questions. (FIXME)"
 ---
-FIXME
+
```
Comment (mr-c) [inline _episodes/01-introduction.md:15]: Add a footnote saying that some of this text is from

> Michael R. Crusoe, Sanne Abeln, Alexandru Iosup, Peter Amstutz, John Chilton, Nebojša Tijanić, Hervé Ménager, Stian Soiland-Reyes, Bogdan Gavrilović, Carole Goble, The CWL Community (2021):
Methods Included: Standardizing Computational Reuse and Portability with the Common Workflow Language.
Communication of the ACM. https://doi.org/10.1145/3486897 Retrieved from arXiv 2105.07028 [cs.DC] https://arxiv.org/abs/2105.07028

File: _episodes/02-shell_to_cwl.md [hunk 2]
```diff
@@ -1,47 +1,513 @@
...
+{: .language-yaml}
+We set the type of `message_text` to string, and set the `inputBinding` `position` (defining where the input item appears after the `baseCommand`) as 1. We also give a default value, `hello World!`, for this item, which will be used if the item is not defined within an input file.
+
+We can now validate, and then run, this tool again:
+~~~
+$ cwl-runner --validate echo.cwl
+~~~
+{: .language-bash}
+~~~
+INFO .../cwl-runner 3.0.20200807132242
```
Comment (mr-c) [inline _episodes/02-shell_to_cwl.md:187]: Maybe remove this line? Or otherwise remove the version number?

File: _episodes/02-shell_to_cwl.md [hunk 3]
```diff
@@ -1,47 +1,513 @@
...
+INFO Resolved 'echo.cwl' to 'file:///.../echo.cwl'
+ERROR Tool definition failed validation:
+echo.cwl:1:1: Object `echo.cwl` is not valid because
+                        tried `CommandLineTool` but
+                          missing required field `inputs`
+~~~
+{: .output}
+`cwl-runner` has noted that we are missing an `inputs` field, so we will add this also to our script (again as an empty list):
```
Comment (mr-c) [inline _episodes/02-shell_to_cwl.md:113]: `s/cwl-runner/cwltool/g`

File: index.md [hunk 4]
```diff
@@ -95,49 +95,23 @@ After following one of these tutorials, learners will be able to:
...
 - __explain how a workflow document describes the input and output of a workflow and the flow of data between tools__ (lesson-level objective)
   - explain the difference between a CWL tool description and a CWL workflow
   - describe the relationship between a tool and its corresponding CWL document
   - exercise good practices when naming inputs and outputs
   - Be able to make understandable and valid names for inputs and outputs (not "input3")
-- __describe all the requirements for running a tool__ (lesson-level objective)
+- __describe the file and variable requirements for running a tool__ (lesson-level objective)
   - identify all the requirements of a tool and define them in the tool description
   - use `runtime` parameters to access information about the runtime environment
   - define environment variables necessary for execution
```
Comment (mr-c) [inline index.md:111]: environment variables might be a better fit for the intermediate lesson

File: _episodes/02-shell_to_cwl.md [hunk 5]
```diff
@@ -1,47 +1,513 @@
...
+# "describing requirements objectives:"
+# "identify all the requirements of a tool and define them in the tool description"
+# "use `runtime` parameters to access information about the runtime environment"
+# "define environment variables necessary for execution"
+# "use `secondaryFiles` or `InitialWorkDirRequirement` to access files in the same directory as another referenced file"
+# "use `$(runtime.cores)` to define the number of cores to be used"
+## "use `type: File`, instead of a string, to reference a filepath"
+- "understand how to reference files for input and output"
+#- "converting shell to cwl objectives:"
+# "identify tasks, and data links in a script"
```
Comment (mr-c) [inline _episodes/02-shell_to_cwl.md:24]: Can we have a companion PR against the [intermediate repo](https://github.com/common-workflow-lab/cwl-intermediate-tutorial) that moves stuff over?

And then you can delete lines instead commenting them out

**Review by mr-c — CHANGES_REQUESTED** [review]
Domain examples (even if we will change them later) are welcome!

**Comment by mr-c** [comment]
@hmenager can you also review?

**Review by mr-c — APPROVED** [review]
Approving and merging so that @aluesink can implement them in follow-up PRs

**Pull request merged**